crates/proof-of-sql/src/sql/proof_plans/range_check.rs

```diff
@@ -0,0 +1,698 @@
+//! Implements a cryptographic range check using logarithmic derivatives to decompose a column of scalars
+//! into a matrix of words. This method leverages the properties of logarithmic derivatives to efficiently
+//! verify range proofs in a zero-knowledge setting by performing word-wise decompositions, intermediate MLEs,
+//! and modular inversions.
+//!
+//! The approach builds on the techniques outlined in the paper "Multivariate Lookups Based on Logarithmic
+//! Derivatives" [ePrint 2022/1530](https://eprint.iacr.org/2022/1530.pdf), which characterizes the use of
+//! logarithmic derivatives to perform multivariate lookups in cryptographic protocols.
+//!
+//! ## Key Steps:
+//! * Word-Sized Decomposition: Each scalar is decomposed into its byte-level representation, forming a matrix where
+//!   each row corresponds to the decomposition of a scalar and each column corresponds to the bytes from the same position
+//!   across all scalars.
+//! * Intermediate MLE Computation: Multi-linear extensions are computed for each word column and for the count of how
+//!   often each word appears.
+//! * Logarithmic Derivative Calculation: After decomposing the scalars, the verifier's challenge is added to each word,
+//!   and the modular multiplicative inverse of this sum is computed, forming a new matrix of logarithmic derivatives.
+//!   This matrix is key to constructing range constraints.
+//!
+//! ## Optimization Opportunities:
+//! * Batch Inversion: Inversions of large vectors are computationally expensive
+//! * Parallelization: Single-threaded execution of these operations is a performance bottleneck
+use crate::{
+    base::{
+        database::TableRef, map::IndexMap, polynomial::MultilinearExtension, scalar::Scalar,
+        slice_ops,
+    },
+    sql::proof::{CountBuilder, FinalRoundBuilder, SumcheckSubpolynomialType, VerificationBuilder},
+};
+use alloc::{boxed::Box, vec::Vec};
+use bumpalo::Bump;
+use bytemuck::cast_slice;
+use core::{cmp::max, iter::repeat};
+
+/// Prove that a word-wise decomposition of a collection of scalars
+/// are all within the range 0 to 2^248. (want this power to be flexible, want to pick word size and range)
+#[allow(clippy::missing_panics_doc)]
+pub fn final_round_evaluate_range_check<'a, S: Scalar + 'a>(
+    builder: &mut FinalRoundBuilder<'a, S>,
+    scalars: &[S],
+    table_length: usize,
+    alloc: &'a Bump,
+) {
+    // Create 31 columns, each will collect the corresponding word from all scalars.
+    // 31 because a scalar will only ever have 248 bits of data set.
+    let mut word_columns: Vec<&mut [u8]> = repeat(())
+        .take(31)
```

Comment:
All occurrences of this value will be replaced by a dynamic size that is determined from a chosen word size